src/mlpack/core/tree/spill_tree/spill_tree.hpp

```diff
@@ -0,0 +1,432 @@
+   * dataset.
+   *
+   * @param data Dataset to create tree from.
+   * @param tau Overlapping size.
+   * @param maxLeafSize Size of each leaf in the tree.
+   * @param rho Balance threshold.
+   */
+  SpillTree(MatType&& data,
+            const double tau = 0,
+            const size_t maxLeafSize = 20,
+            const double rho = 0.7);
+
+  /**
+   * Construct this node as a child of the given parent, starting at column
+   * begin and using count points.  The ordering of that subset of points in the
+   * parent's data matrix will be modified!  This is used for recursive
+   * tree-building by the other constructors which don't specify point indices.
+   *
+   * @param parent Parent of this node.  Its dataset will be modified!
+   * @param points Vector of indexes of points to be included in this node.
+   * @param tau Overlapping size.
+   * @param maxLeafSize Size of each leaf in the tree.
+   * @param rho Balance threshold.
+   */
+  SpillTree(SpillTree* parent,
+            std::vector<size_t>& points,
+            const double tau = 0,
+            const size_t maxLeafSize = 20,
+            const double rho = 0.7);
+
+  /**
+   * Create a hybrid spill tree by copying the other tree.  Be careful!  This
+   * can take a long time and use a lot of memory.
+   *
+   * @param other tree to be replicated.
+   */
+  SpillTree(const SpillTree& other);
+
+  /**
+   * Move constructor for a SpillTree; possess all the members of the
+   * given tree.
+   *
+   * @param other tree to be moved.
+   */
+  SpillTree(SpillTree&& other);
+
+  /**
+   * Initialize the tree from a boost::serialization archive.
+   *
+   * @param ar Archive to load tree from.  Must be an iarchive, not an oarchive.
+   */
+  template<typename Archive>
+  SpillTree(
+      Archive& ar,
+      const typename boost::enable_if<typename Archive::is_loading>::type* = 0);
+
+  /**
+   * Deletes this node, deallocating the memory for the children and calling
+   * their destructors in turn.  This will invalidate any pointers or references
+   * to any nodes which are children of this one.
+   */
+  ~SpillTree();
+
+  //! Return the bound object for this node.
+  const BoundType& Bound() const { return bound; }
+  //! Return the bound object for this node.
+  BoundType& Bound() { return bound; }
+
+  //! Return the statistic object for this node.
+  const StatisticType& Stat() const { return stat; }
+  //! Return the statistic object for this node.
+  StatisticType& Stat() { return stat; }
+
+  //! Return whether or not this node is a leaf (true if it has no children).
+  bool IsLeaf() const;
+
+  //! Gets the left child of this node.
+  SpillTree* Left() const { return left; }
+  //! Modify the left child of this node.
+  SpillTree*& Left() { return left; }
+
+  //! Gets the right child of this node.
+  SpillTree* Right() const { return right; }
+  //! Modify the right child of this node.
+  SpillTree*& Right() { return right; }
```

Comment:
If you prefer, I am happy to leave out the `Left()` and `Right()` functions, and have users access the left and right children as `Child(0)` and `Child(1)`.  I think all of the algorithms that use trees should be depending on `Child()` anyway.